Update Apple TV integration to support tvOS 15

postlund opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This bumps pyatv to the latest version (0.9.7) and makes various adjustments to follow the more generalized API. This mainly concerns the pairing process as that is completely general now, with no special cases where protocol X or Y must be paired or it won't work. Another improvement is that each protocol is verified if pairing or password is needed and provides proper action or error message accordingly.

Change log for pyatv is here: https://github.com/postlund/pyatv/releases/tag/v0.9.0 (0.9.0 has the most important changes, following versions are just minor fixes)

https://github.com/postlund/pyatv/compare/v0.8.2...v0.9.7

The gist of what this update brings:

* Support for tvOS 15 in general
* All device types that pyatv currently supports (Apple TV all generations, HomePods, AirPort Express, iTunes/Music app on macOS and most AirPlay v1 receivers)
* Local streaming of files to the receiver via AirPlay (RAOP)
* AirPlay support for generic receivers is limited, e.g. streaming files is supported, stop button and metadata is updated when Home Assistant is streaming but not when someone else is streaming (limitation in AirPlay)
* I have added a "force re-auth" flag to config flow options to re-trigger set up, which might be needed if a protocol failed to pair the first time (e.g. access control prevented it).

There are a few question marks on my end that I need help with:

* There is no migration path for existing config entries. Old entries must be removed and devices added again. This is simply because the (unique) identifier used to discover the device was part of the MRP protocol, which Apple has removed. So Home Assistant will just keep looking for a device that doesn't exist. I have made this more future proof by saving all identifiers in the config entry, so this will not happen again in the future. How should I handle this now though?
* I would somehow like to raise `ConfigEntryAuthFailed` when authentication fails, but that is only possible in `async_setup_entry`. Since I have a connect loop in a background task, that is not possible. I'm not sure what to do.
* One feature lots of people have requested is TTS support for the HomePod. By supporting audio streaming, that is technically possible but the problem is that only local streaming is supported, i.e. the source must be a local file. The way TTS works today caches the output speech as a file but pass on a URL to `play_media`, so it doesn't work in my case. I did a hack (which obviously should not be in the code) that, when a URL to a TTS-file is passed to `play_media`, extracts the filename and waits for that particular file to appear on disk. This works, but it's not very pretty, so I would prefer a better solution to this. I have a discussion in architecture [here](https://github.com/home-assistant/architecture/discussions/576).

I think that's it for now.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #51649 fixes #54098 fixes #57067
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [X] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [X] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

postlund committed:
Update Apple TV integration to support tvOS 15

postlund committed:
Fix review comments

postlund committed:
Remove TTS hack

MartinHjelmare commented on homeassistant/components/apple_tv/__init__.py:
```diff
@@ -103,26 +87,19 @@ async def async_migrate_entry(hass, config_entry):
 
         config_entry.data = {**new}
-        config_entry.version = 2
+        config_entry.version = 3
```
We shouldn't change config entry version or use the migration api. Just check for missing item in `async_setup_entry` and update the config entry data etc as needed.

postlund replied:
Huh? Is it deprecated? I thought the purpose of it was to ensure version and code matches to avoid having to make optional checks.

MartinHjelmare replied:
As I mentioned above, we want to avoid making rollback impossible. Increasing the config entry version will make rolling back impossible.

postlund replied:
Ok, sure, it wasn't very clear from the documentation. Then I guess I can remove the migrator completely then.

postlund replied:
I changed back to version 1, removed the migration method and added a check if `CONF_IDENTIFIERS` is present in the config entry during setup (and add it if it's not).

postlund committed:
Do not migrate config entries

postlund committed:
Add missing break

bdraco commented:
Just installed for testing. The mdns processing is using ~5x all the other cpu time Home Assistant is using combined.  I'll see if I can spend some time on https://github.com/jstasiak/python-zeroconf/issues/1004 later this week

postlund committed:
Fix handling of identifiers

postlund commented:
> Just installed for testing. The mdns processing is using ~5x all the other cpu time Home Assistant is using combined. I'll see if I can spend some time on [jstasiak/python-zeroconf#1004](https://github.com/jstasiak/python-zeroconf/issues/1004) later this week

This will be most apparent if you have devices in Home Assistant that either aren't on your network or has no power (pyatv is quite eager trying to find a device and traffic will increase linearly with number of devices). The biggest optimization will be continuous scanning that I have planned for some time in the future. In that case the same "session" could be shared by all devices, I.e each device would not have to scan but could depend on the shared instance that scans in the background.

MartinHjelmare commented on homeassistant/components/apple_tv/config_flow.py:
```diff
@@ -181,7 +214,7 @@ async def async_find_device_wrapper(self, next_func, allow_exist=False):
             return self.async_abort(reason="no_devices_found")
         except DeviceAlreadyConfigured:
-            return self.async_abort(reason="already_configured")
+            return self.async_abort(reason="already_configured_device")
```
```suggestion
            return self.async_abort(reason="already_configured")
```

MartinHjelmare commented on homeassistant/components/apple_tv/strings.json:
```diff
@@ -45,17 +52,21 @@
       "already_configured_device": "[%key:common::config_flow::abort::already_configured_device%]",
```
```suggestion
      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]",
```

MartinHjelmare commented on homeassistant/components/apple_tv/strings.json:
```diff
@@ -45,17 +52,21 @@
         "data": {
-          "start_off": "Do not turn device on when starting Home Assistant"
+          "start_off": "Do not turn device on when starting Home Assistant",
+          "reconfigure": "Force reconfiguration of device"
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/apple_tv/strings.json:
```diff
@@ -45,17 +52,21 @@
         "description": "Configure general device settings",
         "data": {
-          "start_off": "Do not turn device on when starting Home Assistant"
+          "start_off": "Do not turn device on when starting Home Assistant",
```
```suggestion
          "start_off": "Do not turn device on when starting Home Assistant"
```

MartinHjelmare commented on homeassistant/components/apple_tv/manifest.json:
```diff
@@ -3,8 +3,15 @@
+    "_appletv-v2._tcp.local.",
+    "_airplay._tcp.local.",
+    "_raop._tcp.local.",
+    "_hscp._tcp.local."
```
Do we need the new ids for this PR?

postlund replied:
Separate PRs for each service?

MartinHjelmare replied:
No, just one PR that adds new services.

This PR is about upgrading to a new library version and making only the changes needed to still have a working integration. All other changes should go in other PRs to keep the scope as small as possible.

postlund replied:
I removed all the newly added services, so only the old ones are left. Fixed remaining of your suggestions.

bdraco replied:
I just tested this and the integration can't find apple tvs that have upgraded to TvOS 15 as the two original services that this was looking for are no longer being announced

bdraco replied:
Listening for `_airplay` without an additional filter is problematic though as it matches way too many things.

bdraco replied:
`{"type":"_airplay._tcp.local.","model":"AppleTV*"}` should pickup AppleTVs


`{"type":"_airplay._tcp.local.","model":"AudioAccessory*"}` should pickup HomePods

bdraco replied:
That should be the most stable AFAICT since apple is least likely to alter the airplay spec as it needs to be interoperable with other vendors

bdraco replied:
Looks like we can only limit by `manufacturer` and not `model` in `zeroconf`

bdraco replied:
..or if we prefer airplay, I'll add support for `model` to `zeroconf`

postlund replied:
You are right, it won't work. Only the already added services as well as airplay are necessary as neither RAOP nor Companion is currently used by anything (technically RAOP is used by `play_media` but that isn't a problem). We must filter the model as you are suggesting though, only Apple TVs should be found (not HomePods) if we really should stick with just the uplift with no features. Seems unnecessary to spend too much time with this though as I will open a new PR immediately after this being merged, adding the missing services. Processes in all glory, sometimes it's better to just be pragmatic.

bdraco replied:
https://github.com/home-assistant/core/pull/58922

postlund replied:
@bdraco Once merged, how do I use that? I guess we only want to match "AppleTV*" for now.

postlund replied:
Right, I see how that works now.

postlund committed:
Fix more review comments

MartinHjelmare commented on homeassistant/components/apple_tv/__init__.py:
```diff
@@ -229,7 +225,12 @@ async def _connect_loop(self):
             except exceptions.AuthenticationError:
-                self._auth_problem()
+                self.config_entry.async_start_reauth(self.hass)
+                asyncio.create_task(self.disconnect())
```
Should we make sure that we are disconnected before we start the re-auth flow to avoid multiple connections at the same time? Does it matter?

postlund replied:
Doesn't matter, connection will closed long before re-auth starts and having multiple connections open is not a problem either.

MartinHjelmare commented on homeassistant/components/apple_tv/config_flow.py:
```diff
@@ -87,88 +69,139 @@ def async_get_options_flow(config_entry):
-                errors["base"] = "already_configured"
-            except exceptions.NoServiceError:
-                errors["base"] = "no_usable_service"
+                errors["base"] = "already_configured_device"
```
```suggestion
                errors["base"] = "already_configured"
```

MartinHjelmare commented:
Should we include this in the release today?

MartinHjelmare commented:
Marking it for the milestone so we don't forget about this.

MartinHjelmare commented:
Please resolve the merge conflicts.

postlund commented:
> Should we include this in the release today?



> Should we include this in the release today?

Would be great!

postlund commented:
> Please resolve the merge conflicts.

I have fixed that locally, it's just the matter of the model filter by @bdraco (it's not merged yet). What should I put in the manifest?

frenck commented:
> it's just the matter of the model filter by @bdraco (it's not merged yet). 

That is not in yet, nor is it part of the beta. We are close to releasing it. I think this changeset it too large to include last minute. However, that said, it will fix the integration for a lot of users. So that is a bit of an unfortunate choice.

Let's start by resolving the merge conflict; as long as that conflict is here, it for sure won't be in the release.

Additionally, can this work without further changes to the manifest?

postlund commented:
> That is not in yet, nor is it part of the beta. We are close to releasing 2021.11, so the time left is short (only a few hours... at max!). I think this changeset it too large to include last minute. However, that said, it will fix the integration for a lot of users. So that is a bit of an unfortunate choice.
> 
> Let's start by resolving the merge conflict; as long as that conflict is here, it for sure won't be in the release.
> 
> Additionally, can this work without further changes to the manifest?

It's a catch-22 because current changes to the manifest will not work as AirPlay devices are not found. Since tvOS 15 MRP (used for remote control support) is tunneled over AirPlay, which means that it must be added to the manifest. This however means other AirPlay devices are discovered as well. So the filter provided by @bdraco is needed to just filter out Apple TV devices. Otherwise we need a hack in the config flow that does more or less the same thing.

frenck commented:
I've removed this change from the milestone, lets test this in dev/beta first.

frenck commented:
@postlund #58922 has been merged, maybe it's time to move this PR forward to ensure we get it all in, in time for 2021.12?

postlund commented:
> @postlund #58922 has been merged, maybe it's time to move this PR forward to ensure we get it all in, in time for 2021.12?

Great! I will fix it ASASP!

postlund committed:
Fix more review comments

postlund commented:
Last comments should be fixed now I hope. Unfortunately I don't have any way to verify that it actually works right now, so if anyone (maybe @bdraco) can test the PR before merging, that would be great!

bdraco commented:
<img width="389" alt="Screen Shot 2021-11-14 at 2 46 58 PM" src="https://user-images.githubusercontent.com/663432/141698062-d07fb693-a08d-45cc-adaf-38b6edbd3163.png">
<img width="290" alt="Screen Shot 2021-11-14 at 2 46 03 PM" src="https://user-images.githubusercontent.com/663432/141698063-d4de4c23-1ae2-41ad-bf94-cb199e469a81.png">

Discovery worked. I was able to add it and do TTS

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented:
Please write the breaking change paragraph so we can copy it directly to the release notes.

postlund commented:
Thanks for the fast response here 👍 I have updated the breaking change text, I hope it's good enough!

MartinHjelmare commented:
It's not clear where to remove the devices. Is that in Home Assistant or somewhere else? If it's in Home Assistant, shouldn't we do that automatically for the user?

postlund commented:
From Home Assistant is what I'm referring to. The reason is described in the initial post:

*There is no migration path for existing config entries. Old entries must be removed and devices added again. This is simply because the (unique) identifier used to discover the device was part of the MRP protocol, which Apple has removed. So Home Assistant will just keep looking for a device that doesn't exist. I have made this more future proof by saving all identifiers in the config entry, so this will not happen again in the future. How should I handle this now though?*

If it is possible to remove it, then we probably should. But I don't know how to do that properly and AFAIK there's no developer documentation describing how to do this or even if it's a practice to do so.

MartinHjelmare commented:
I think we should re-configure all config entries that don't have the stored identifiers. I commented about that here:
https://github.com/home-assistant/core/pull/58665#discussion_r739010570

MartinHjelmare reviewed (changes requested):
There still seems to be work left to do.

postlund commented:
The problem is that we can't reconfigure as we don't have any information to match a config entry to its corresponding device on the network. The information is disjunctive. I'm up for suggestions to solve it.

MartinHjelmare commented:
We can ask the user for any information we need in the config flow. How would we do it for a new device? If we can configure new devices we should be able to solve this. If we can't find a matching device in the end of the flow, we can remove the config entry and let the user know.

postlund commented:
Ok, I will see if I can change things around a bit and make that happen. I think it deviates a bit from how it works right now, so it probably needs some work.

postlund committed:
Bump to pyatv 0.9.7

postlund commented:
Took some time to come around with this because of life. But I have made some updates to pyatv and fixed so the MRP identifier discovered again, so no breaking change needs to happen. An integration must be removed and re-added to pair new protocols though since it's not possible to manually trigger re-configuration. I still believe this is something that Home assistant core should offer (not for further discussion here).

I have also implemented initial support for audio streaming from HTTP(S) in pyatv and made a minor tweak, which means TTS works without any hacks now.

Would be great if someone could test this, in particular with a config entry from a device running an official release to make sure device discovery works properly.

bdraco commented:
> I still believe this is something that Home assistant core should offer (not for further discussion here).

https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#reauthentication

I think you can use the `reauthentication` workflow

postlund committed:
Fix zeroconf mock in tests

postlund commented:
> > I still believe this is something that Home assistant core should offer (not for further discussion here).
> 
> https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#reauthentication
> 
> I think you can use the `reauthentication` workflow

Yes, I can probably come up with something to deal with that. It will have to be another PR though (same problem exists today).

MartinHjelmare commented:
There are some test failures. Should we remove the breaking change label and paragraph now?

postlund commented:
> There are some test failures. Should we remove the breaking change label and paragraph now?

I apparent failed at pushing the fix for that. Will hopefully succeed now. I removed the breaking change paragraph as well.

MartinHjelmare reviewed (approved):
Thanks!

MartinHjelmare commented:
There is no github release for version 0.9.7:
https://github.com/postlund/pyatv/releases

postlund commented:
> There is no github release for version 0.9.7:
> https://github.com/postlund/pyatv/releases

Right, had some Actions issues, but published it manually now.

postlund commented:
Woho! 🎉 Will it be picked for 2021.12.0 or wait until the next after that?

postlund commented:
Sure, that's fine by me! That gives me time to fix some more stuff before then.
